File: Backend/DataAccess/BillFlow.Models/Entities/CompanySettings.cs
namespace BillFlow.Models.Entities;

public class CompanySettings
{
    public Guid OwnerId { get; set; }

    public User Owner { get; set; } = null!;

    public string CompanyName { get; set; } = null!;

    public string? Address { get; set; }

    public string? Country { get; set; }

    public string? TaxNumber { get; set; }

    public string? PhoneNumber { get; set; }

    public string? Email { get; set; }

    public string Currency { get; set; } = "USD";

    public string InvoiceNumberPrefix { get; set; } = "INV";

    public decimal DefaultTaxRate { get; set; }

    public int PaymentTermsDays { get; set; } = 30;

    public string? TimeZone { get; set; }

    public string? BrandColor { get; set; }

    public string? InvoiceFooterNote { get; set; }

    public byte[]? LogoBytes { get; set; }

    public string? LogoContentType { get; set; }

    public bool EnablePaymentReminders { get; set; }

    public int ReminderDaysBeforeDue { get; set; } = 3;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime? UpdatedAt { get; set; }
}

File: Frontend/src/domain/billing/company-settings.ts
export type CompanySettingsResponse = {
  companyName: string;
  address?: string | null;
  country?: string | null;
  taxNumber?: string | null;
  phoneNumber?: string | null;
  email?: string | null;
  currency: string;
  invoiceNumberPrefix: string;
  defaultTaxRate: number;
  paymentTermsDays: number;
  timeZone?: string | null;
  brandColor?: string | null;
  invoiceFooterNote?: string | null;
  hasLogo?: boolean;
  enablePaymentReminders?: boolean;
  reminderDaysBeforeDue?: number;
  createdAt: string;
  updatedAt?: string | null;
};

export type UpsertCompanySettingsRequest = {
  companyName: string;
  address?: string;
  country?: string;
  taxNumber?: string;
  phoneNumber?: string;
  email?: string;
  currency: string;
  invoiceNumberPrefix: string;
  defaultTaxRate: number;
  paymentTermsDays: number;
  timeZone?: string;
  brandColor?: string;
  invoiceFooterNote?: string;
  enablePaymentReminders?: boolean;
  reminderDaysBeforeDue?: number;
};

export const defaultCompanySettingsForm = {
  companyName: "",
  address: "",
  country: "",
  taxNumber: "",
  phoneNumber: "",
  email: "",
  currency: "USD",
  invoiceNumberPrefix: "INV",
  defaultTaxRate: 0,
  paymentTermsDays: 30,
  timeZone: "",
  brandColor: "#FF6B00",
  invoiceFooterNote: "",
  enablePaymentReminders: false,
  reminderDaysBeforeDue: 3,
};

export function mapSettingsToForm(
  settings: CompanySettingsResponse,
): typeof defaultCompanySettingsForm {
  return {
    companyName: settings.companyName,
    address: settings.address ?? "",
    country: settings.country ?? "",
    taxNumber: settings.taxNumber ?? "",
    phoneNumber: settings.phoneNumber ?? "",
    email: settings.email ?? "",
    currency: settings.currency,
    invoiceNumberPrefix: settings.invoiceNumberPrefix,
    defaultTaxRate: settings.defaultTaxRate,
    paymentTermsDays: settings.paymentTermsDays,
    timeZone: settings.timeZone ?? "",
    brandColor: settings.brandColor ?? "#FF6B00",
    invoiceFooterNote: settings.invoiceFooterNote ?? "",
    enablePaymentReminders: settings.enablePaymentReminders ?? false,
    reminderDaysBeforeDue: settings.reminderDaysBeforeDue ?? 3,
  };
}

export function mapFormToRequest(
  values: typeof defaultCompanySettingsForm,
): UpsertCompanySettingsRequest {
  const optional = (value: string) => {
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : undefined;
  };

  return {
    companyName: values.companyName.trim(),
    address: optional(values.address),
    country: optional(values.country),
    taxNumber: optional(values.taxNumber),
    phoneNumber: optional(values.phoneNumber),
    email: optional(values.email),
    currency: values.currency.trim().toUpperCase(),
    invoiceNumberPrefix: values.invoiceNumberPrefix.trim(),
    defaultTaxRate: values.defaultTaxRate,
    paymentTermsDays: values.paymentTermsDays,
    timeZone: optional(values.timeZone),
    brandColor: optional(values.brandColor),
    invoiceFooterNote: optional(values.invoiceFooterNote),
    enablePaymentReminders: values.enablePaymentReminders,
    reminderDaysBeforeDue: values.reminderDaysBeforeDue,
  };
}
